### PR TITLE
Add `localizedPriceString` to `StoreProductDiscount`

### PR DESCRIPTION
**Pull request by joshdholtz**
### Motivation
- Need by hybrid common migration to v4
  - https://github.com/RevenueCat/purchases-hybrid-common/pull/113/files#diff-2135910c11a83bffe485630bb393f14e8451d1b8069b6f2439ba0abe4ab18695L50
- Not easy way to get localized price string for discounts (that aren't an intro)

### Description
- Added `localizedPriceString` to `StoreProductDiscountType`

### CODE REVIEW

File: APITesters/SwiftAPITester/SwiftAPITester/StoreProductAPI.swift [hunk 1]
```diff
@@ -8,6 +8,8 @@
 import RevenueCat
 
 var product: StoreProduct!
+
+// swiftlint:disable:next function_body_length
```
Comment (NachoSoto) [inline APITesters/SwiftAPITester/SwiftAPITester/StoreProductAPI.swift:12]: Maybe we could disable these for the APITester folders?

File: StoreKitUnitTests/StoreProductTests.swift [hunk 2]
```diff
@@ -215,7 +215,7 @@ class StoreProductTests: StoreKitConfigTestCase {
         var priceFormatter = try XCTUnwrap(storeProduct.priceFormatter)
         var productPrice = storeProduct.price as NSNumber
 
-        expect(priceFormatter.string(from: productPrice)) == "4,99 €"
+        expect(priceFormatter.string(from: productPrice)) == "€4.99"
```
Comment (NachoSoto) [inline StoreKitUnitTests/StoreProductTests.swift:218]: This is wrong now 😕 that's not how prices are formatted in Euro. I guess the locale is wrong?

File: Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift [hunk 3]
```diff
@@ -131,7 +131,7 @@ internal protocol StoreProductType {
     /// - SeeAlso: ``pricePerMonth``.
     var price: Decimal { get }
 
-    /// The price of this product using ``priceFormatter``.
+    /// The price of this product formatted for locale.
```
Comment (NachoSoto) [inline Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift:134]: Why this change?

File: Purchases/Purchasing/StoreKitAbstractions/StoreProductDiscount.swift [hunk 4]
```diff
@@ -70,7 +70,9 @@ public final class StoreProductDiscount: NSObject, StoreProductDiscountType {
     @objc public var offerIdentifier: String? { self.discount.offerIdentifier }
     @objc public var currencyCode: String? { self.discount.currencyCode }
     @objc public var price: Decimal { self.discount.price }
+    @objc public var localizedPriceString: String { self.discount.localizedPriceString}
```
Comment (NachoSoto) [inline Purchases/Purchasing/StoreKitAbstractions/StoreProductDiscount.swift:73]: Nit: missing space: 
```suggestion
    @objc public var localizedPriceString: String { self.discount.localizedPriceString }
```

**Review by NachoSoto — CHANGES_REQUESTED** [review]
Overall looks good, like I said consistency might be worth it. I like that this is defined in the protocol now. But unfortunately the test change shows this has broken the previous behavior 😕

File: Purchases/Purchasing/StoreKitAbstractions/StoreProductDiscount.swift (hunk 4 above)
Comment (NachoSoto) [inline Purchases/Purchasing/StoreKitAbstractions/StoreProductDiscount.swift:73]: This needs to go to the APITesters too I think?

File: Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift (hunk 3 above)
Comment (joshdholtz) [reply Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift:134]: Only SK1 uses `priceFormatter` for `localizedPriceString`. SK2 uses the `displayPrice` so the comment was wrong 🤷‍♂️  Should we be using `priceFormatter` for both SK1 and SK2?

File: Purchases/Purchasing/StoreKitAbstractions/StoreProductDiscount.swift (hunk 4 above)
Comment (joshdholtz) [reply Purchases/Purchasing/StoreKitAbstractions/StoreProductDiscount.swift:73]: Maybe I should write code in API testers first and _then_ write the actual code so I don't forget. API Tester driven development?

File: APITesters/SwiftAPITester/SwiftAPITester/StoreProductAPI.swift (hunk 1 above)
Comment (joshdholtz) [reply APITesters/SwiftAPITester/SwiftAPITester/StoreProductAPI.swift:12]: How do we do that? 😇

File: StoreKitUnitTests/StoreProductTests.swift (hunk 2 above)
Comment (joshdholtz) [reply StoreKitUnitTests/StoreProductTests.swift:218]: Ohhh... `formatter.locale = .autoupdatingCurrent` isn't affected by `SKTestSession` but it instead uses the locale set in the test scheme 🤦‍♂️ 

So... going to have mock the locale somehow 🤷‍♂️

Comment (NachoSoto) [reply StoreKitUnitTests/StoreProductTests.swift]: Well no, the test is correct, this would fail in real code too. I think it's because we need to use the actual locale in the product after all.

File: StoreKitUnitTests/StoreProductTests.swift [hunk 5]
```diff
@@ -212,7 +212,11 @@ class StoreProductTests: StoreKitConfigTestCase {
 
         var storeProduct = try await sk1Fetcher.product(withIdentifier: Self.productID)
 
+        // This formatter would normally get locale form user preferences
+        // but that's not possible to change in the tests so manually setting
         var priceFormatter = try XCTUnwrap(storeProduct.priceFormatter)
+        priceFormatter.locale = Locale(identifier: "es_ES")
```
Comment (NachoSoto) [inline StoreKitUnitTests/StoreProductTests.swift:218]: This is just making the test pass but the implementation is wrong :/

File: StoreKitUnitTests/StoreProductTests.swift (hunk 2 above)
Comment (joshdholtz) [reply StoreKitUnitTests/StoreProductTests.swift]: I'm super confused now 😔 We can only get the locale from SK1 products (which is the SK1 tests were testing for `4,99 €`). The SK2 tests can't get the locale so they using `autoupdatingCurrent` which were returning `€4.99`

But doesn't my setting of the locale in the tests essentially mock the `.autoupdatingCurrent`? In the US I'll get `€4.99` but in Spain I'll get `4,99 €`?

But there is another thing to consider... Only SK1 is using `priceFormatter` for `localizedPriceString`. SK2 is using the `displayPrice` from the SK2 object (which whatever format Apple returns). This may also lead to some inconsistency 😬 

Do we want `localizedPriceString` and `priceFormatter` to return the same for SK1 and SK2 or use the StoreKit default?

File: APITesters/SwiftAPITester/SwiftAPITester/StoreProductAPI.swift (hunk 1 above)
Comment (NachoSoto) [reply APITesters/SwiftAPITester/SwiftAPITester/StoreProductAPI.swift:12]: You could create a `.swiftlint.yml` (like `PurchasesTests/.swiftlint.yml`) inside of `APITester/` with this rule, then I think it'll disable it for all testers since it's not a rule we care about for these files :) But we can clean this up after this PR.

**Review by NachoSoto — APPROVED** [review]
🚢 it

File: APITesters/SwiftAPITester/SwiftAPITester/StoreProductAPI.swift (hunk 1 above)
Comment (NachoSoto) [reply APITesters/SwiftAPITester/SwiftAPITester/StoreProductAPI.swift:12]: Done: #1309

**Comment by alfredcc** [comment]
So how can I get `currencySymbol` in `StoreProductDiscount`?

**Comment by NachoSoto** [comment]
I looked into this and unfortunately it doesn't seem like `Product.SubscriptionOffer` exposes this directly or through `Product.jsonRepresentation`.
What's your use case that `localizedPriceString` doesn't cover?

**Comment by alfredcc** [comment]
> I looked into this and unfortunately it doesn't seem like `Product.SubscriptionOffer` exposes this directly or through `Product.jsonRepresentation`. What's your use case that `localizedPriceString` doesn't cover?

Sorry for the late response. I found that we no longer need it, and `localizedPriceString` can cover the case.

**Comment by aboedo** [comment]
@alfredcc thanks for getting back to us! I'm glad `localizedPriceString` works for your use case 👍
